### PR TITLE
feat(prune): don't prune changesets if taking an incremental hashing/trie route

### PR DESCRIPTION
**Pull request by joshieDo**
summary:
* execution stage checks hashing and merkle stages clean thresholds before following the user prune configuration for changesets. If any stage will take an incremental approach, then don't prune any changesets. this will happen during pipeline re-runs and they will be cleaned up eventually by `Pruner`.
* removes any prune logic from merkle and hashing stages.

### CODE REVIEW

File: crates/stages/src/stages/execution.rs [hunk 1]
```diff
@@ -163,6 +181,33 @@ impl<EF: ExecutorFactory> ExecutionStage<EF> {
...
+    /// This function verifies whether the MerkleStage or Hashing stages will run from scratch.
+    /// If at least one stage isn't starting anew, it implies that pruning of changesets cannot
+    /// occur. This is determined by checking the highest clean threshold
+    /// (`self.external_clean_threshold`) across the stages.
+    ///
+    /// Given that `start_block` changes with each checkpoint, it's necessary to inspect
+    /// [`tables::AccountsTrie`] and [`tables::HashedAccount`] to ensure that the stages haven't
+    /// been previously executed.
+    fn adjust_prune_modes<DB: Database>(
+        &mut self,
```
[the remark this answers is not in the document]
Comment (joshieDo) [reply crates/stages/src/stages/execution.rs]: it wasn't actually modifying the value, it was just taking a copy. but no need for mut yea

File: crates/stages/src/stages/execution.rs [hunk 2]
```diff
@@ -163,6 +181,33 @@ impl<EF: ExecutorFactory> ExecutionStage<EF> {
...
+        provider: &DatabaseProviderRW<'_, &DB>,
+        start_block: u64,
+        max_block: u64,
+    ) -> Result<PruneModes, StageError> {
+        let mut prune_modes = self.prune_modes;
+        if !(max_block - start_block > self.external_clean_threshold ||
+            (provider.tx_ref().cursor_read::<tables::AccountsTrie>()?.first()?.is_none() &&
+                provider.tx_ref().cursor_read::<tables::HashedAccount>()?.first()?.is_none()))
```
Comment (shekhirin) [inline crates/stages/src/stages/execution.rs:204]: ultranit: i think `provider.tx_ref().entries::<T>()? == 0` will be cheaper here, as it doesn't require initializing a cursor and going through database

**Review by shekhirin — APPROVED** [review]
LGTM, makes sense to me. It feels a bit hacky, but I think that's fine

File: crates/stages/src/stages/execution.rs (hunk 2 above)
Comment (joshieDo) [reply crates/stages/src/stages/execution.rs:204]: i had that initially, but i was unsure on how it works underneath

**Comment by shekhirin** [comment]
DO NOT MERGE (yet): syncing Sepolia right now, will post results here

**Comment by joshieDo** [comment]
sepolia passed fine

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**